src/types/decimal.rs

```diff
@@ -296,8 +296,42 @@ impl TryFrom<f64> for Decimal {
 
 impl Display for Decimal {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        // TODO: This is correct, but not the most human-friendly format.
-        write!(f, "{}d{}", self.coefficient, self.exponent)
+        let digits = &*self.coefficient.magnitude.to_string();
+        let len = digits.len();
+        // The index of the decimal, relative to the magnitude representation
+        //       0123
+        // Given ABCDd-2, the decimal gets inserted at position 2.
+        let i_d = len as i64 + self.exponent;
+
+        if self.coefficient.sign == Sign::Negative {
+            write!(f, "-").unwrap();
+        };
+
+        if self.exponent == 0 && len > 6 {
+            // e.g. A.BCDEFGd6
+            write!(f, "{}.{}d{}", &digits[0..1], &digits[1..len], (i_d - 1))
+        } else if self.exponent == 0 {
+            // e.g. ABC.
+            write!(f, "{}.", &digits)
+        } else if self.exponent >= 0 {
+            // e.g. ABCd1
+            // let zeroes = "0".repeat(self.exponent as usize);
+            write!(f, "{}d{}", &digits, self.exponent)
+        } else {
+            // exponent < 0, there is a fractional component
+            if i_d > 0 {
+                // e.g. A.BC or AB.C
+                let ui_d = i_d as usize;
```

Comment:
If it's helpful, you can shadow the existing name in this scope so you don't have to think of a new one. (IMO, `ui_d` is a bit cryptic, though I gather it stands for `unsigned_index_of_d`?). Shadowing has the potential to be a bit confusing for readers outside of an IDE, but in this case it only lasts one line.